src/daemon/rpmostreed-transaction-types.cxx

```diff
@@ -2781,35 +2781,76 @@ kernel_arg_transaction_finalize (GObject *object)
-  g_autofree char **append_if_missing = static_cast<char **>(vardict_lookup_strv_canonical (self->options, "append-if-missing"));
-  g_autofree char **delete_if_present = static_cast<char **>(vardict_lookup_strv_canonical (self->options, "delete-if-present"));
-  gboolean changed = FALSE;
+  g_return_val_if_fail (self != NULL, FALSE);
+  g_return_val_if_fail (upgrader != NULL, FALSE);
+  g_return_val_if_fail (kargs != NULL, FALSE);
 
-  /* don't want to pull new content for this */
-  upgrader_flags |= RPMOSTREE_SYSROOT_UPGRADER_FLAGS_SYNTHETIC_PULL;
-  upgrader_flags |= RPMOSTREE_SYSROOT_UPGRADER_FLAGS_PKGCACHE_ONLY;
-  if (vardict_lookup_bool (self->options, "lock-finalization", FALSE))
-    upgrader_flags |= RPMOSTREE_SYSROOT_UPGRADER_FLAGS_LOCK_FINALIZATION;
+  g_auto(GStrv) kargs_strv = ostree_kernel_args_to_strv (kargs);
+  rpmostree_sysroot_upgrader_set_kargs (upgrader, kargs_strv);
 
-  /* Read in the existing kernel args and convert those to an #OstreeKernelArg instance for API usage */
-  g_autoptr(OstreeKernelArgs) kargs = ostree_kernel_args_from_string (self->existing_kernel_args);
-  g_autoptr(RpmOstreeSysrootUpgrader) upgrader =
-    rpmostree_sysroot_upgrader_new (sysroot, self->osname, static_cast<RpmOstreeSysrootUpgraderFlags>(upgrader_flags),
-                                    cancellable, error);
-  if (upgrader == NULL)
+  if (!rpmostree_sysroot_upgrader_deploy (upgrader, NULL, cancellable, error))
     return FALSE;
-  rpmostree_sysroot_upgrader_set_caller_info (upgrader, command_line, 
-                                              rpmostreed_transaction_get_agent_id (RPMOSTREED_TRANSACTION(self)),
-                                              rpmostreed_transaction_get_sd_unit (RPMOSTREED_TRANSACTION(self)));
+
+  if (vardict_lookup_bool (self->options, "reboot", FALSE))
+    {
+      if (!check_sd_inhibitor_locks (cancellable, error))
+        return FALSE;
+      rpmostreed_daemon_reboot (rpmostreed_daemon_get ());
+    }
+
+  return TRUE;
+}
+
+static gboolean
+kernel_arg_apply_final_str(KernelArgTransaction *self,
+                           RpmOstreeSysrootUpgrader *upgrader,
+                           const char *final_kernel_args,
+                           GCancellable *cancellable,
+                           GError **error)
+{
+  g_return_val_if_fail (self != NULL, FALSE);
+  g_return_val_if_fail (upgrader != NULL, FALSE);
+  g_return_val_if_fail (final_kernel_args != NULL, FALSE);
+  g_return_val_if_fail (self->existing_kernel_args != NULL, FALSE);
+
+  gboolean changed = (strcmp (self->existing_kernel_args, final_kernel_args) != 0);
+
+  if (changed)
+    {
+      g_autoptr(OstreeKernelArgs) kargs = ostree_kernel_args_from_string (final_kernel_args);
+      if (!kernel_arg_apply (self, upgrader, kargs, cancellable, error))
+        return FALSE;
+    }
+  else
+    {
+      rpmostree_output_message ("No changes.");
+    }
```

Comment:
Seems like we could fold change detection into `kernel_arg_apply` instead of having it in both `final_str` and `patching`, no?

Reply:
I switched to that, to be honest both ways look a bit silly but for different reasons :)